hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/ReadableAccountStore.java

```diff
@@ -68,25 +70,61 @@ public KeyOrLookupFailureReason getKey(final AccountID idOrAlias) {
         if (account.isEmpty()) {
             return withFailureReason(INVALID_ACCOUNT_ID);
         }
-        return validateKey(account.get().getAccountKey());
+        return validateKey(account.get().getAccountKey(), false);
     }
 
     /** {@inheritDoc} */
     @Override
-    public KeyOrLookupFailureReason getKeyIfReceiverSigRequired(final AccountID idOrAlias) {
+    public KeyOrLookupFailureReason getKeyIfReceiverSigRequired(
+            @NonNull final AccountID idOrAlias) {
         final var account = getAccountLeaf(idOrAlias);
         if (account.isEmpty()) {
             return withFailureReason(INVALID_ACCOUNT_ID);
         }
 
-        final var responseIgnoringSigReq = validateKey(account.get().getAccountKey());
+        final var responseIgnoringSigReq = validateKey(account.get().getAccountKey(), false);
         if (responseIgnoringSigReq.failed() || account.get().isReceiverSigRequired()) {
             return responseIgnoringSigReq;
         } else {
             return PRESENT_BUT_NOT_REQUIRED;
         }
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public KeyOrLookupFailureReason getKey(@NonNull final ContractID idOrAlias) {
+        final var optContract = getContractLeaf(idOrAlias);
+        if (optContract.isEmpty()) {
+            return withFailureReason(INVALID_CONTRACT_ID);
+        }
+        final var contract = optContract.get();
+        if (contract.isDeleted() || !contract.isSmartContract()) {
+            return withFailureReason(INVALID_CONTRACT_ID);
+        }
+        return validateKey(contract.getAccountKey(), true);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public KeyOrLookupFailureReason getKeyIfReceiverSigRequired(
+            @NonNull final ContractID idOrAlias) {
+        final var optContract = getContractLeaf(idOrAlias);
+        if (optContract.isEmpty()) {
+            return withFailureReason(INVALID_CONTRACT_ID);
+        }
+
+        final var contract = optContract.get();
+        if (contract.isDeleted() || !contract.isSmartContract()) {
+            return withFailureReason(INVALID_CONTRACT_ID);
+        }
```

Comment:
Duplicated in `getKey(...)`. Not a huge deal since it's a small code snippet. Hopefully it doesn't come back to haunt us 😅

Reply:
Moved the duplicated 2 checks to a different method. Thank you!